test/ng/directive/formSpec.js

```diff
@@ -58,6 +58,114 @@ describe('form', function() {
+      '</form>')(scope);
+
+    var form = scope.myForm;
+
+    var input = doc.find('input').eq(0);
+    var inputController = input.controller('ngModel');
+
+    changeInputValue(input, 'ab');
+    scope.$apply();
+
+    expect(form.$error.maxlength).toBeTruthy();
+    expect(form.$dirty).toBe(true);
+    expect(form.$error.maxlength[0].$name).toBe('control');
+
+    // remove control
+    form.$removeControl(form.control);
+    expect(form.control).toBeUndefined();
+    expect(form.$error.maxlength).toBeFalsy();
+
+    inputController.$setPristine();
+    expect(form.$dirty).toBe(true);
+
+    form.$setPristine();
+
+    changeInputValue(input, 'abc');
+    scope.$apply();
+
+    expect(form.$error.maxlength).toBeFalsy();
+    expect(form.$dirty).toBe(false);
+  });
+
+
+  it('should react to changes in manually added controls', function() {
+    doc = $compile(
+      '<form name="myForm">' +
+        '<input name="control" ng-maxlength="1" ng-model="value" store-model-ctrl/>' +
+      '</form>')(scope);
+
+    var form = scope.myForm;
+
+    var input = doc.find('input').eq(0);
+    var inputController = input.controller('ngModel');
+
+    // remove control so we can add it again
+    form.$removeControl(form.control);
+    //invalidate the control
+    changeInputValue(input, 'abc');
+
+    expect(inputController.$error.maxlength).toBe(true);
+    expect(inputController.$dirty).toBe(true);
+    expect(form.control).toBeUndefined();
+    expect(form.$error.maxlength).toBeFalsy();
+    expect(form.$dirty).toBe(false);
+
+    //re-add the control
+    form.$addControl(inputController);
+    changeInputValue(input, 'a');
+
+    expect(form.$error.maxlength).toBeFalsy();
+  });
+
+
+  it('should use the correct parent when renaming and removing dynamically added controls', function() {
+    scope.controlName = 'childControl';
+    scope.hasChildControl = true;
+
+    doc = $compile(
+      '<form name="myForm">' +
+        '<div ng-if="hasChildControl">' +
+          '<input name="{{controlName}}" ng-maxlength="1" ng-model="value"/>' +
+        '</div>' +
+      '</form>' +
+      '<form name="otherForm"></form>')(scope);
+
+    scope.$digest();
+
+    var form = scope.myForm;
+    var otherForm = scope.otherForm;
+    var childControl = form.childControl;
+
+    // remove child form and add it to another form
+    form.$removeControl(childControl);
+    otherForm.$addControl(childControl);
+
+    expect(form.childControl).toBeUndefined();
+    expect(otherForm.childControl).toBe(childControl);
+
+    // rename the childControl
+    scope.controlName = 'childControlMoved';
+    scope.$digest();
+
+    expect(form.childControlMoved).toBeUndefined();
+    expect(otherForm.childControl).toBeUndefined();
+    expect(otherForm.childControlMoved).toBe(childControl);
+
+    scope.hasChildControl = false;
+    scope.$digest();
+
+    expect(form.childControlMoved).toBeUndefined();
+    expect(otherForm.childControlMoved).toBeUndefined();
```

Comment:
Sorry - I was just thinking loud (laughing actually).

I find it funny (in a good way), that you can have a control be one form's child (in DOM), but belong to another form (in Angular terms).
I thought it was the wrong expectation at first site, but it is correct and find it nice that it works.

Sorry for the noise and confusion :disappointed_relieved: